File: DsaWithCpp/LinkedList/SinglyLinkedList.cpp
#include <iostream>
#include "Node.cpp"
using namespace std;

class SinglyLinkedList
{
    Node *head = NULL;
    int size = 0;

public:
    bool isEmpty()
    {
        return head == NULL;
    }

    void insertAtBegin(int data)
    {
        Node *newNode = new Node(data);
        if (isEmpty())
        {
            head = newNode;
            cout << "Data Inserted!\n";
            size++;
            return;
        }
        newNode->next = head;
        head = newNode;
        cout << "Data Inserted!\n";
        size++;
    }

    void insertAtPos(int data, int pos)
    {
        if (pos == 0)
            cout << "Invalid Input! Index Starts from 1.\n";
        else if (pos == 1)
            insertAtBegin(data);
        else if (pos > size)
            cout << "Invalid Input! List Size is: " << size << endl;
        else
        {
            Node *newNode = new Node(data);
            Node *currentNode = head;
            for (int i = 0; i < pos - 2; i++)
                currentNode = currentNode->next;
            newNode->next = currentNode->next;
            currentNode->next = newNode;
            cout << "Data Inserted!\n";
            size++;
        }
    }
    void insertAtLast(int data)
    {
        Node *newNode = new Node(data);
        if (isEmpty())
        {
            head = newNode;
            cout << "Data Inserted!\n";
            size++;
            return;
        }
        Node *currentNode = head;
        while (currentNode->next != NULL)
            currentNode = currentNode->next;
        currentNode->next = newNode;
        cout << "Data Inserted!\n";
        size++;
    }

    void remAtBegin()
    {
        if (isEmpty())
            cout << "List Is Empty!\n";
        else if (head->next == NULL)
        {
            head = NULL;
            cout << "List Successfully Cleared!\n";
            size = 0;
        }
        else
        {
            head = head->next;
            cout << "Successfully Deleted!\n";
            size--;
        }
    }

    void remAtPos(int pos)
    {
        if (isEmpty())
            cout << "List Is Empty!\n";
        else if (head->next == NULL)
        {
            head = NULL;
            cout << "List Successfully Cleared!\n";
            size = 0;
        }
        else if (pos = 0)
            cout << "Invalid Input! Index Starts from 1.\n";
        else if (pos == 1)
            remAtBegin();
        else if (pos > size)
            cout << "Invalid Input! List Size is: " << size << endl;
        else
        {
            Node *currentNode = head;
            for (int i = 0; i < pos - 2; i++)
                currentNode = currentNode->next;
            currentNode->next = currentNode->next->next;
            cout << "Sucessfully Deleted!\n";
            size--;
        }
    }

    void remAtLast()
    {
        if (isEmpty())
            cout << "List Is Empty!\n";
        else if (head->next == NULL)
        {
            head = NULL;
            cout << "List Successfully Cleared!\n";
            size = 0;
        }
        else
        {
            Node *lastNode = head;
            while (lastNode->next->next != NULL)
                lastNode = lastNode->next;
            lastNode->next = NULL;
            cout << "Successfully Deleted!\n";
            size--;
        }
    }

    void removeAll()
    {
        if (isEmpty())
        {
            cout << "List Is Empty!\n";
            return;
        }
        cout << "List Successfully Cleared!\n";
        head = NULL;
        size = 0;
    }

    void display()
    {
        if (isEmpty())
        {
            cout << "List is Empty!\n";
            return;
        }
        Node *currentNode = head;
        printf("[");
        while (currentNode != NULL)
        {
            printf("%i, ", currentNode->data);
            currentNode = currentNode->next;
        }
        printf("\b\b]\nList Size is -> %i\n", size);
    }
};

File: Java/linkedlist/SinglyLinkedList.java
package linkedlist;

public class SinglyLinkedList {
    int size = 0;
    Node head;

    public boolean isEmpty() {
        return head == null;
    }

    public void insertAtBegin(int data) {
        Node newNode = new Node(data);
        // If list is Empty
        if (isEmpty()) {
            head = newNode;
            System.out.println("Data Inserted!");
            size++;
            return;
        }
        // newNode next will point to firstNode
        newNode.next = head;
        // head is set to newNode
        head = newNode;
        System.out.println("Data Inserted!");
        size++;
    }

    public void insertAtPos(int data, int pos) {
        // If index is 0
        if (pos == 0)
            System.out.println("Invalid Input! Index Starts from 1.");
        // If index is 1
        else if (pos == 1)
            insertAtBegin(data);
        // If index is greater than the list size
        else if (pos > size)
            System.out.println("Invalid Input! List Size is: " + size);
        else {
            Node newNode = new Node(data);
            Node currentNode = head;
            // traverse till index
            for (int i = 0; i < pos - 2; i++)
                currentNode = currentNode.next;
            // this assign the next of currentNode i.e.
            // address of next element to the next of newNode
            newNode.next = currentNode.next;
            // this assign the newNode to the next of currentNode
            // i.e. address of newNode is assigned in the next of currentNode
            currentNode.next = newNode;
            System.out.println("Data Inserted!");
            size++;
        }
    }

    public void insertAtLast(int data) {
        Node newNode = new Node(data);
        // If list is Empty
        if (isEmpty()) {
            head = newNode;
            System.out.println("Data Inserted!");
            size++;
            return;
        }
        Node currentNode = head;
        // traverse till last element
        while (currentNode.next != null)
            currentNode = currentNode.next;
        // lastNode next is set to newNode
        currentNode.next = newNode;
        System.out.println("Data Inserted!");
        size++;
    }

    public void delAtBegin() {
        // If List is Empty
        if (isEmpty())
            System.out.println("List is Empty!");
        // If list have one element
        else if (head.next == null) {
            head = null;
            System.out.println("List Successfully Cleared!");
            size = 0;
        } else {
            // head from the firstNode is set to secNode, this removes the firstNode
            head = head.next;
            System.out.println("Successfully Deleted!");
            size--;
        }
    }

    public void delAtPos(int pos) {
        // If List is Empty
        if (isEmpty())
            System.out.println("List is Empty!");
        // If list has only 1 element
        else if (head.next == null) {
            head = null;
            System.out.println("List Successfully Cleared!");
            size = 0;
        }
        // If index is 0
        else if (pos == 0)
            System.out.println("Invalid Input! Index Starts from 1.");
        else if (pos == 1)
            delAtBegin();
        // If index is equal to list size
        else if (pos == size)
            delAtLast();
        // If index is greater than the list size
        else if (pos > size)
            System.out.println("Invalid Input! List Size is: " + size);
        else {
            Node currentNode = head;
            // traverse till index
            for (int j = 0; j < pos - 2; j++)
                currentNode = currentNode.next;
            // this assigns the address of 2nd next element to the currentNode,
            // to the next of currentNode, removing the mid element
            currentNode.next = currentNode.next.next;
            System.out.println("Successfully Deleted!");
            size--;
        }
    }

    public void delAtLast() {
        // If List is Empty
        if (isEmpty())
            System.out.println("List is Empty!");
        // If list have one element
        else if (head.next == null) {
            head = null;
            System.out.println("List Successfully Cleared!");
            size = 0;
        } else {
            Node lastNode = head;
            // traverse till second last element
            while (lastNode.next.next != null)
                lastNode = lastNode.next;
            // second Last Node is set to null, this removes the address,
            // removing the last element
            lastNode.next = null;
            System.out.println("Successfully Deleted!");
            size--;
        }
    }

    public void deleteAll() {
        // If List is Empty
        if (isEmpty()) {
            System.out.println("List is Empty!");
            return;
        }
        System.out.println("List Successfully Cleared!");
        head = null;
        size = 0;
    }

    public void display() {
        // If list is Empty
        if (isEmpty()) {
            System.out.println("List is Empty!");
            return;
        }
        Node currentNode = head;
        System.out.print("[");
        while (currentNode != null) {
            System.out.print(currentNode.data + ", ");
            currentNode = currentNode.next;
        }
        System.out.println("\b\b]\nList Size -> " + size);
    }
}
